File: CS/WindowsApplication3/SelectionHelper.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;
using DevExpress.XtraEditors;
using DevExpress.XtraVerticalGrid;
using System.Diagnostics;
using System.Collections;
using DevExpress.XtraVerticalGrid.Rows;
using System.Linq;
using DevExpress.LookAndFeel;

namespace DXSample
{
    public class SelectionHelper
    {
        private SelectOperaion _SelectionOperation;
        private Selection _Selection;
        VGridControlBase vgrid;

        public SelectionHelper(VGridControlBase vgrid)
        {
            this.vgrid = vgrid;
            Selection = new Selection(vgrid);
        }

        public Selection Selection
        {
            get { return _Selection; }
            set
            {
                if (_Selection != value)
                {
                    _Selection = value;
                }
            }
        }

        private SelectOperaion SelectionOperation
        {
            get {
                if (_SelectionOperation == null)
                    _SelectionOperation = new SelectOperaion(Selection);
                return _SelectionOperation;
            }
        }

        public void SelectCell(GridCell cell)
        {
            Selection.StartRangeCell = null;
            if (Selection.Contains(cell))
                Selection.Remove(cell);
            else
                Selection.Add(cell);
        }


        public bool IsEmpty
        {
            get { return Selection.Count == 0; }
        }

        public void SelectCellRange(GridCell endCell)
        {
            GridCell startCell = Selection[Selection.Count - 1];
            if (Selection.StartRangeCell != null)
            {
                int index = Selection.IndexOf(Selection.StartRangeCell);
                while (Selection.Count > index + 1)
                    Selection.RemoveAt(Selection.Count - 1);
                startCell = Selection.StartRangeCell;
            }
            else
                Selection.StartRangeCell = startCell;

            int startRow = Math.Min(startCell.Row.VisibleIndex, endCell.Row.VisibleIndex);
            int endRow = Math.Max(startCell.Row.VisibleIndex, endCell.Row.VisibleIndex);
            int startRecord = Math.Min(startCell.Record, endCell.Record);
            int endRecord = Math.Max(startCell.Record, endCell.Record);
            SelectionOperation.StartCell = new Point(startRow, startRecord);
            SelectionOperation.EndCell = new Point(endRow, endRecord);
            vgrid.RowsIterator.DoOperation(SelectionOperation);
        }

        public void ClearAndSelectCell(GridCell cell)
        {
            Selection.StartRangeCell = null;
            Selection.Clear();
            Selection.Add(cell);
        }

        public void Clear()
        {
            Selection.Clear();
        }

        public bool Contains(GridCell cell)
        {
            return Selection.Contains(cell);
        }

        public void SelectAll()
        {
            SelectionOperation.StartCell = new Point(0, 0);
            SelectionOperation.EndCell = new Point(vgrid.Rows.VisibleRowsCount, vgrid.RecordCount);
            vgrid.RowsIterator.DoOperation(SelectionOperation);
        }

        public void RefreshSelection()
        {
            foreach (GridCell cell in Selection)
                vgrid.InvalidateRow(cell.Row);
        }
    }
}

File: VB/WindowsApplication3/SelectionHelper.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Text
Imports System.Windows.Forms
Imports DevExpress.XtraEditors
Imports DevExpress.XtraVerticalGrid
Imports System.Diagnostics
Imports System.Collections
Imports DevExpress.XtraVerticalGrid.Rows
Imports System.Linq
Imports DevExpress.LookAndFeel

Namespace DXSample
	Public Class SelectionHelper
		Private _SelectionOperation As SelectOperaion
		Private _Selection As Selection
		Private vgrid As VGridControlBase

		Public Sub New(ByVal vgrid As VGridControlBase)
			Me.vgrid = vgrid
			Selection = New Selection(vgrid)
		End Sub

		Public Property Selection() As Selection
			Get
				Return _Selection
			End Get
			Set(ByVal value As Selection)
				If _Selection IsNot value Then
					_Selection = value
				End If
			End Set
		End Property

		Private ReadOnly Property SelectionOperation() As SelectOperaion
			Get
				If _SelectionOperation Is Nothing Then
					_SelectionOperation = New SelectOperaion(Selection)
				End If
				Return _SelectionOperation
			End Get
		End Property

		Public Sub SelectCell(ByVal cell As GridCell)
			Selection.StartRangeCell = Nothing
			If Selection.Contains(cell) Then
				Selection.Remove(cell)
			Else
				Selection.Add(cell)
			End If
		End Sub


		Public ReadOnly Property IsEmpty() As Boolean
			Get
				Return Selection.Count = 0
			End Get
		End Property

		Public Sub SelectCellRange(ByVal endCell As GridCell)
			Dim startCell As GridCell = Selection(Selection.Count - 1)
			If Selection.StartRangeCell IsNot Nothing Then
				Dim index As Integer = Selection.IndexOf(Selection.StartRangeCell)
				Do While Selection.Count > index + 1
					Selection.RemoveAt(Selection.Count - 1)
				Loop
				startCell = Selection.StartRangeCell
			Else
				Selection.StartRangeCell = startCell
			End If

			Dim startRow As Integer = Math.Min(startCell.Row.VisibleIndex, endCell.Row.VisibleIndex)
			Dim endRow As Integer = Math.Max(startCell.Row.VisibleIndex, endCell.Row.VisibleIndex)
			Dim startRecord As Integer = Math.Min(startCell.Record, endCell.Record)
			Dim endRecord As Integer = Math.Max(startCell.Record, endCell.Record)
			SelectionOperation.StartCell = New Point(startRow, startRecord)
			SelectionOperation.EndCell = New Point(endRow, endRecord)
			vgrid.RowsIterator.DoOperation(SelectionOperation)
		End Sub

		Public Sub ClearAndSelectCell(ByVal cell As GridCell)
			Selection.StartRangeCell = Nothing
			Selection.Clear()
			Selection.Add(cell)
		End Sub

		Public Sub Clear()
			Selection.Clear()
		End Sub

		Public Function Contains(ByVal cell As GridCell) As Boolean
			Return Selection.Contains(cell)
		End Function

		Public Sub SelectAll()
			SelectionOperation.StartCell = New Point(0, 0)
			SelectionOperation.EndCell = New Point(vgrid.Rows.VisibleRowsCount, vgrid.RecordCount)
			vgrid.RowsIterator.DoOperation(SelectionOperation)
		End Sub

		Public Sub RefreshSelection()
			For Each cell As GridCell In Selection
				vgrid.InvalidateRow(cell.Row)
			Next cell
		End Sub
	End Class
End Namespace